layers/descriptor_sets.cpp

```
@@ -0,0 +1,942 @@
+            auto immut = p_layout_->GetImmutableSamplerPtrFromIndex(i);
+            for (uint32_t di = 0; di < p_layout_->GetDescriptorCountFromIndex(i); ++di) {
+                if (immut)
+                    descriptors_.push_back(std::unique_ptr<Descriptor>(new ImageSamplerDescriptor(
+                        immut + di, image_view_map_, image_map_, image_to_swapchain_map_, swapchain_map_, sampler_map_)));
+                else
+                    descriptors_.push_back(std::unique_ptr<Descriptor>(new ImageSamplerDescriptor(
+                        image_view_map_, image_map_, image_to_swapchain_map_, swapchain_map_, sampler_map_)));
+            }
+            break;
+        }
+        // ImageDescriptors
+        case VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE:
+        case VK_DESCRIPTOR_TYPE_INPUT_ATTACHMENT:
+        case VK_DESCRIPTOR_TYPE_STORAGE_IMAGE:
+            for (uint32_t di = 0; di < p_layout_->GetDescriptorCountFromIndex(i); ++di)
+                descriptors_.push_back(std::unique_ptr<Descriptor>(
+                    new ImageDescriptor(type, image_view_map_, image_map_, image_to_swapchain_map_, swapchain_map_)));
+            break;
+        case VK_DESCRIPTOR_TYPE_UNIFORM_TEXEL_BUFFER:
+        case VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER:
+            for (uint32_t di = 0; di < p_layout_->GetDescriptorCountFromIndex(i); ++di)
+                descriptors_.push_back(std::unique_ptr<Descriptor>(new TexelDescriptor(type, buffer_view_map_)));
+            break;
+        case VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER:
+        case VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC:
+        case VK_DESCRIPTOR_TYPE_STORAGE_BUFFER:
+        case VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC:
+            for (uint32_t di = 0; di < p_layout_->GetDescriptorCountFromIndex(i); ++di)
+                descriptors_.push_back(std::unique_ptr<Descriptor>(new BufferDescriptor(type, buffer_map_)));
+            break;
+        default:
+            break;
```

Silent default case here is worrying -- if it were ever hit, all subsequent indices get quietly misaligned.